runtime/vm/xz64/UpcallThunkGen.cpp

```diff
@@ -0,0 +1,767 @@
+	thunkMem += LG(thunkMem, 14, 0, C_SP_RegisterNum, 112);
+	thunkMem += BASR(thunkMem, 14, 14);
+
+	// Finish up before returning
+	vmFuncs->doneUpcallThunkGeneration(metaData, (void *)(metaData->thunkAddress));
+	// printf("-- createUpCallThunk -- thunkAddress: %p\n", (uint64_t *)metaData->thunkAddress);
+	return (void *)(metaData->thunkAddress);
+}
+
+/**
+ * @brief Calculate the requested argument in-stack memory address to return
+ * @param nativeSig[in] A pointer to the J9UpcallNativeSignature
+ * @param argListPtr[in] A pointer to the argument list prepared by the thunk
+ * @param argIdx[in] The requested argument index
+ * @return address in argument list for the requested argument
+ *
+ * Details:
+ *   A quick walk-through of the argument list ahead of the requested one
+ *   Calculating its address based on argListPtr
+ */
+void *
+getArgPointer(J9UpcallNativeSignature *nativeSig, void *argListPtr, I_32 argIdx)
+{
+	J9UpcallSigType *sigArray = nativeSig->sigArray;
+    I_32 lastSigIdx = nativeSig->numSigs - 1; // The index for the return type in the signature array
+
+	Assert_VM_true((argIdx >= 0) && (argIdx < lastSigIdx));
+	bool isPointerToStruct = false;
+
+	// Test the return type
+	I_32 structOffset = 0;
+	I_32 structSize = sigArray[argIdx].sizeInByte;
+
+	I_32 gprIndex = 1;
+	I_32 fprIndex = -2;
+	I_32 callerFrameIndex = -1; // values that weren't passed by register
+	I_32 calleeFrameIndex = -1; // values that were passed through register
+	bool argumentInCallerFrame = false; // At first we assume the argument hasn't overflow and is in the new frame allocated by the callee
+	I_32 paramAreaOffset = 160; // offset from the start of a stack frame to start of parameter area holding overflow arguments
+
+	switch (sigArray[lastSigIdx].type & J9_FFI_UPCALL_SIG_TYPE_MASK) {
+		// If the return type is a struct, then the first entry in the calleeFrame is an address to a buffer
+		// where the struct will be stored when we return to native caller. So we must skip over that
+		// "hiddenParameter" when trying to identify arguments in getArgPointer.
+		case J9_FFI_UPCALL_SIG_TYPE_STRUCT:
+			calleeFrameIndex++;
+			break;
+		default:
+			break;
+	}
+
+	switch(sigArray[argIdx].type & J9_FFI_UPCALL_SIG_TYPE_MASK) {
+		case J9_FFI_UPCALL_SIG_TYPE_STRUCT:
+			// structs not of size = 1,2,4,8 are passed by pointers. These must be dereferenced before
+            // returning. So we set a flag here to do so.
+			if (structSize != 1 && structSize != 2 && structSize != 4 && structSize != 8) {
+					isPointerToStruct = true;
+			}
+			else
+			{
+					// structs that are passed by value are expected to be stored as right aligned in the 8 byte
+					// stack slot. The return address to the struct must point to the start of the struct. So for
+					// structs smaller than 8 that are passed by value we must increment the offset by the required amount
+					// to adjust the final return address.
+					structOffset = 8 - structSize;
```

Comment:
mentioned above:  you cannot handle FPs similarly as to integer arguments.

Reply:
for ALL_SP with a single float, this calculation is obviously having structOffset set to 4.

Reply:
Ah I see what you mean. It does look like I missed that corner case. I'll rework this when I push new changes with the formatting fixed.

I'll also be adding to the upcall tests to capture all the zLinux corner cases (either in this PR or I'll pass them on to Cheng) so hopefully we won't have any more of these.